alg_convex_hull.go

```diff
@@ -0,0 +1,119 @@
+package simplefeatures
+
+import (
+	"sort"
+)
+
+const (
+	// Right indicates the orientation is right turn which is anticlockwise
+	Right = iota
+	// Collinear indicates three points are on the same line
+	Collinear
+	// Left indicates the orientation is left turn which is clockwise
+	Left
+)
+
+// graphamScan returns a convex hull.
+func grahamScan(ps []XY) []XY {
+	if len(ps) < 3 {
+		return nil
+	}
+	sortedPoints := sortByPolarAngle(ps)
+
+	s := make([]XY, 2, len(sortedPoints))
+	copy(s, sortedPoints[:2])
+	t := make([]XY, len(sortedPoints)-2)
+	copy(t, sortedPoints[2:])
+
+	for i := 0; i < len(t); i++ {
+		ori := orientation(s[len(s)-2], s[len(s)-1], t[i])
+		switch {
+		case ori == Left:
+			s = append(s, t[i])
+		default:
+			s = s[:len(s)-1]
+			s = append(s, t[i])
+		}
+	}
+
+	return append(s, s[0])
+}
+
+// soryByPolarAngle sorts the points by their polar angle
+func sortByPolarAngle(ps []XY) []XY {
```

Comment:
Since `sortByPolarAngle` works in-place, I don't think we need to return anything.